File: docs-vitepress/scripts/preview_server.py
#!/usr/bin/env python3
"""UncivCN 文档站本地预览服务器。

- 服务 docs-vitepress/.vitepress/dist，映射 /Unciv/ 前缀（与 GitHub Pages 路径一致）
- 无扩展名路径自动补 .html（如 /zh/UncivCN/新特性 -> 新特性.html）
- 空闲超时（默认 5 分钟无 HTTP 访问）自动退出，Ctrl+C 立即退出
- 关闭浏览器后无需手动清理进程

用法：python scripts/preview_server.py [--port 4173] [--idle-timeout 300]
"""
import argparse
import http.server
import os
import sys
import threading
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent  # docs-vitepress/
DIST = ROOT / '.vitepress' / 'dist'

BASE = '/Unciv/'
last_access = time.time()


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(DIST), **kwargs)

    def translate_path(self, path: str) -> str:
        # /Unciv/xxx -> /xxx（去掉 base 前缀）
        if path == BASE.rstrip('/') or path == BASE:
            path = '/'
        elif path.startswith(BASE):
            path = path[len(BASE) - 1:]  # 保留前导 /
        return super().translate_path(path)

    def _serve(self):
        # 无扩展名的路径自动尝试 .html（保留 query）
        if '?' in self.path:
            base, query = self.path.split('?', 1)
        else:
            base, query = self.path, ''
        p = self.translate_path(base)
        if not os.path.exists(p) and not base.endswith('/') and '.' not in base.rsplit('/', 1)[-1]:
            self.path = base + '.html' + (('?' + query) if query else '')
        super().do_GET()

    def do_GET(self):
        self._serve()

    def do_HEAD(self):
        self._serve()

    def log_message(self, fmt: str, *args):
        global last_access
        last_access = time.time()
        print(f"[{time.strftime('%H:%M:%S')}] {self.address_string()} {fmt % args}", flush=True)


def main() -> int:
    global last_access
    parser = argparse.ArgumentParser(description='UncivCN 文档站本地预览服务器')
    parser.add_argument('--port', type=int, default=4173)
    parser.add_argument('--idle-timeout', type=int, default=300,
                        help='无 HTTP 访问多少秒后自动退出（默认 300）')
    args = parser.parse_args()

    if not DIST.exists():
        print(f'[错误] 未找到构建产物 {DIST}，请先运行 build.bat 构建。')
        return 1

    try:
        server = http.server.ThreadingHTTPServer(('127.0.0.1', args.port), Handler)
    except OSError as e:
        print(f'[错误] 端口 {args.port} 启动失败：{e}')
        print('      可能是已有预览服务器在运行（直接访问 http://localhost:4173/Unciv/ 即可）。')
        return 1

    print(f'预览地址: http://localhost:{args.port}/Unciv/')
    print(f'空闲 {args.idle_timeout} 秒无访问将自动退出；Ctrl+C 立即停止。')
    print('关闭本窗口或浏览器后无需手动清理进程。')

    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    try:
        while True:
            time.sleep(2)
            if time.time() - last_access > args.idle_timeout:
                print('[提示] 已空闲超时，自动退出。')
                break
    except KeyboardInterrupt:
        print('\n[提示] 已手动停止。')
    finally:
        server.shutdown()
        server.server_close()
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: docs-vitepress/scripts/preview-server.mjs
#!/usr/bin/env node
/**
 * UncivCN 文档站本地预览服务器（Node 版，替代原 preview_server.py）。
 *
 * - 服务 docs-vitepress/.vitepress/dist，映射 /Unciv/ 前缀（与 GitHub Pages 路径一致）
 * - 无扩展名路径自动补 .html（如 /zh/UncivCN/新特性 -> 新特性.html）
 * - 空闲超时（默认 5 分钟无 HTTP 访问）自动退出，Ctrl+C 立即退出
 * - 关闭浏览器后无需手动清理进程
 *
 * 用法：node scripts/preview-server.mjs [--port 4173] [--idle-timeout 300]
 */
import http from 'node:http'
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const DIST = path.resolve(__dirname, '../.vitepress/dist')
const BASE = '/Unciv/'

const args = process.argv.slice(2)
const port = Number(args[args.indexOf('--port') + 1] ?? 4173)
const idleTimeout = Number(args[args.indexOf('--idle-timeout') + 1] ?? 300)

let lastAccess = Date.now()

const MIME = {
  '.html': 'text/html; charset=utf-8', '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8', '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8', '.map': 'application/json',
  '.png': 'image/png', '.svg': 'image/svg+xml', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
  '.gif': 'image/gif', '.webp': 'image/webp', '.ico': 'image/x-icon',
  '.woff2': 'font/woff2', '.txt': 'text/plain; charset=utf-8', '.xml': 'text/xml; charset=utf-8',
}

/** 去掉 /Unciv/ 前缀并解析为 DIST 内的文件路径；越界（../）返回 null */
function resolveFile(urlPath) {
  let p = urlPath
  if (p === BASE || p === BASE.slice(0, -1)) p = '/'
  else if (p.startsWith(BASE)) p = '/' + p.slice(BASE.length)
  let decoded
  try { decoded = decodeURIComponent(p) } catch { return null }
  const full = path.normalize(path.join(DIST, decoded))
  if (full !== DIST && !full.startsWith(DIST + path.sep)) return null
  return full
}

const server = http.createServer((req, res) => {
  lastAccess = Date.now()
  const url = new URL(req.url ?? '/', 'http://localhost')

  let filePath = resolveFile(url.pathname)
  if (filePath === null) { res.writeHead(403).end('Forbidden'); return }

  // 无扩展名的路径自动尝试 .html（如 /zh/UncivCN/新特性）
  if (!fs.existsSync(filePath) && !url.pathname.endsWith('/') &&
      !path.basename(url.pathname).includes('.')) {
    filePath = resolveFile(url.pathname + '.html') ?? filePath
  }
  // 目录请求 -> index.html
  if (fs.existsSync(filePath) && fs.statSync(filePath).isDirectory()) {
    filePath = path.join(filePath, 'index.html')
  }
  if (!fs.existsSync(filePath)) { res.writeHead(404).end('Not Found'); return }

  const ext = path.extname(filePath).toLowerCase()
  res.writeHead(200, { 'Content-Type': MIME[ext] ?? 'application/octet-stream' })
  fs.createReadStream(filePath).pipe(res)
})

server.on('error', (e) => {
  if (e.code === 'EADDRINUSE') {
    console.error(`[错误] 端口 ${port} 启动失败：可能已有预览服务器在运行（直接访问 http://localhost:${port}/Unciv/ 即可）。`)
    process.exit(1)
  }
  console.error(e)
  process.exit(1)
})

server.listen(port, '127.0.0.1', () => {
  console.log(`预览地址: http://localhost:${port}/Unciv/`)
  console.log(`空闲 ${idleTimeout} 秒无访问将自动退出；Ctrl+C 立即停止。`)
  console.log('关闭本窗口或浏览器后无需手动清理进程。')
})

// 空闲超时自动退出
setInterval(() => {
  if (Date.now() - lastAccess > idleTimeout * 1000) {
    console.log('[提示] 已空闲超时，自动退出。')
    process.exit(0)
  }
}, 2000)
